solidity/contracts/midas/WithdrawalQueue.sol

```diff
@@ -0,0 +1,289 @@
+        }
+
+        midasAllocator = MidasAllocator(_midasAllocator);
+    }
+
+    function requestRedeem(
+        uint256 _shares,
+        address _receiver,
+        uint256 _exitFeeInAssets
+    ) external onlyAcreBTC {
+        uint256 tbtcAmount = acrebtc.convertToAssets(_shares);
+        uint256 midasShares = vault.convertToShares(tbtcAmount);
+        midasAllocator.withdraw(midasShares);
+        vaultSharesToken.approve(address(vault), midasShares);
+        acrebtc.burn(_shares);
+        if (_exitFeeInAssets > 0) {
+            uint256 exitFeeShares = vault.convertToShares(_exitFeeInAssets);
+            vault.requestRedeem(exitFeeShares, acrebtc.treasury());
+            midasShares -= exitFeeShares;
+        }
+        vault.requestRedeem(midasShares, _receiver);
+    }
+
+    /// @notice Requests a redemption with extra bridge data
+    /// @param _shares Amount of shares to withdraw.
+    /// @param _walletPubKeyHash Wallet public key hash.
+    /// @param _redeemer The original redeemer address.
+    /// @return requestId The ID of the withdrawal request.
+    function requestRedeemAndBridge(
+        uint256 _shares,
+        bytes20 _walletPubKeyHash,
+        address _redeemer
+    ) external onlyAcreBTC returns (uint256 requestId) {
+        uint256 tbtcAmount = acrebtc.convertToAssets(_shares);
+        uint256 midasShares = vault.convertToShares(tbtcAmount);
+        midasAllocator.withdraw(midasShares);
+        acrebtc.burn(_shares);
+        vaultSharesToken.approve(address(vault), midasShares);
+        uint256 midasRequestId = vault.requestRedeem(
+            midasShares,
+            address(this)
+        );
+
+        requestId = count;
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            redeemer: _redeemer,
+            shares: midasShares,
+            tbtcAmount: tbtcAmount,
+            createdAt: block.timestamp,
+            completedAt: 0,
+            isCompleted: false,
+            walletPubKeyHash: _walletPubKeyHash,
+            midasRequestId: midasRequestId
+        });
+
+        emit WithdrawalRequestCreated(
+            requestId,
+            _redeemer,
+            midasShares,
+            tbtcAmount,
+            _walletPubKeyHash,
+            midasRequestId
+        );
+        count++;
+    }
+
+    /// @notice Completes a withdrawal request.
+    /// @param _requestId ID of the withdrawal request.
+    /// @param _tbtcRedemptionData Additional data required for the tBTC redemption.
+    ///        See `redemptionData` parameter description of `Bridge.requestRedemption`
+    ///        function.
+    function completeWithdrawalRequest(
+        uint256 _requestId,
+        bytes calldata _tbtcRedemptionData
+    ) external onlyMaintainer {
+        // TBTC Token contract owner resolves to the TBTCVault contract.
+        if (tbtc.owner() != tbtcVault) revert UnexpectedTbtcTokenOwner();
+        if (withdrawalRequests[_requestId].isCompleted)
```

Comment:
Instead of introducing the `isCompleted` property we could check if `completedAt` is set.
 
```suggestion
        if (withdrawalRequests[_requestId].completedAt > 0)
```